In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [4]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path('Resources/lending_data.csv')
lending_data_df = pd.read_csv(file_path)
# Review the DataFrame
# Display the first few rows of the DataFrame
lending_data_df.head(10)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  
5                 1       20600            0  
6                 1       21100            0  
7                 0       15100            0  
8                 0       17400            0  
9                 0       18800            0

In [5]:
# display more details about the DataFrame
print(lending_data_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB
None


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [6]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.drop(columns=['loan_status'])

In [10]:
# Review the y variable Series
print("y variable (loan_status):")
y.head()

y variable (loan_status):


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [11]:
# Review the X variable DataFrame
print("\nX DataFrame (features):")
X.head()


X DataFrame (features):


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [12]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# print for review 
print("\nX_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (62028, 7)
X_test shape: (15508, 7)
y_train shape: (62028,)
y_test shape: (15508,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [21]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [22]:
# Make a prediction using the testing data
y_pred = logistic_regression_model.predict(X_test)
print("\nPredictions on the testing data:")
print(y_pred)


Predictions on the testing data:
[0 0 0 ... 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [24]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:")
print(conf_matrix)


Confusion Matrix:
[[14924    77]
 [   31   476]]


In [25]:
# Print the classification report for the model
class_report = classification_report(y_test, y_pred)
print("\nClassification Report:")
print(class_report)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**
To evaluate how well the logistic regression model predicts both the 0 (healthy loan) and 1 (high-risk loan) labels, 
you can examine the confusion matrix and classification report generated in the previous steps. These tools provide detailed insights into the model's performance for each class.

## Confusion Matrix:
The confusion matrix provides a summary of prediction results. It shows the number of true positives, true negatives, false positives, and false negatives.
Confusion Matrix:
[[14924    77]
 [   31   476]]
 
this layout  allows visualization of the performance of an algorithm, typically for supervised learning. 
Each row of the matrix represents the instances in an actual class while each column represents the instances in a predicted class,
 or vice versa. Here's a more detailed breakdown of the confusion matrix components:

[[True Negatives (TN), False Positives (FP)],
 [False Negatives (FN), True Positives (TP)]]

### Interpretation:
True Negatives (TN): The number of healthy loans (label 0) that were correctly predicted as healthy.
False Positives (FP): The number of healthy loans (label 0) that were incorrectly predicted as high-risk (label 1).
False Negatives (FN): The number of high-risk loans (label 1) that were incorrectly predicted as healthy (label 0).
True Positives (TP): The number of high-risk loans (label 1) that were correctly predicted as high-risk.

The matrix results will represent for each parameters like these:
True Negatives (TN): 14924 
False Positives (FP): 77
False Negatives (FN): 31
True Positives (TP): 476

So we can interpret these results as follows:
14,924 loans were correctly predicted as healthy (0).
77 loans were incorrectly predicted as high-risk (1) but are actually healthy (0).
31 loans were incorrectly predicted as healthy (0) but are actually high-risk (1).
476 loans were correctly predicted as high-risk (1).

## Classification Report:
The classification report provides precision, recall, f1-score, and support for each class.
Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508

### Interpretation:
* For Label 0 (Healthy Loan):
    * Precision: 1.00
      This means that out of all the loans predicted as healthy, 100% were actually healthy. The model has an excellent precision for healthy loans, indicating very few false positives.
    * Recall: 0.99
      This means that out of all the actual healthy loans, 99% were correctly identified by the model. The model has a very high recall for healthy loans, indicating very few false negatives.
    * F1-Score: 1.00
      The F1-score, which is the harmonic mean of precision and recall, is 1.00, indicating excellent overall performance for predicting healthy loans.
    * Support: 15,001
      There were 15,001 actual healthy loans in the test dataset.

* For Label 1 (High-Risk Loan):
    * Precision: 0.86
      This means that out of all the loans predicted as high-risk, 86% were actually high-risk. The model has good precision for high-risk loans, but there is some room for improvement.
    * Recall: 0.94
      This means that out of all the actual high-risk loans, 94% were correctly identified by the model. The model has a very high recall for high-risk loans, indicating it captures most of the high-risk loans.
    * F1-Score: 0.90
      The F1-score for high-risk loans is 0.90, indicating good overall performance, though not as perfect as for healthy loans.
    * Support: 507
      There were 507 actual high-risk loans in the test dataset.
### Overall Performance:
* The overall accuracy of the model is 99%, indicating that 99% of the predictions made by the model are correct.
* These are the unweighted averages of the precision, recall, and F1-score for both classes. They provide a general sense of performance but do not account for class imbalance.
* These are the averages weighted by the number of instances in each class. They give a better sense of the model's performance when there is a class imbalance, as in this case.

## Conclusion:
* The logistic regression model performs exceptionally well in predicting healthy loans (label 0) with perfect precision and near-perfect recall.
* The model also performs well in predicting high-risk loans (label 1), with good precision and very high recall, though there is some room for improvement in reducing false positives.
* Overall, the model is highly accurate, and its performance metrics indicate it is reliable for classifying loan statuses.


---